File: 112-path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        int curr=0;

        return dfs(root,targetSum, curr);
        
    }

    bool dfs(TreeNode*root, int targetSum, int curr)
    {
        if(!root) return false;

        curr+=root->val;

        if(!root->left && !root->right)
        {
            return curr==targetSum;

        }

        return dfs(root->left, targetSum, curr) || dfs(root->right, targetSum, curr);
    }
};

File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:

        # so basically in this question you have to check if the path sum adds up to the target or not
        # each node has it's own copy of target

        # so the trick is khudka karo and aage badho

        def dfs(node,remaining)->bool:
            if not node:
                return False
            if not node.left and not node.right and remaining==node.val:
                return True
            print(remaining)
            return dfs(node.left, remaining-node.val) or dfs(node.right, remaining-node.val)
        return dfs(root,targetSum)

       
        